File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
            int ans=0,lft=0,max_cnt=0;
            unordered_map<char,int>help_map;
            for(int rgt=0;rgt<s.size();rgt++){
                help_map[s[rgt]]++;
                max_cnt=max(max_cnt,help_map[s[rgt]]);
                while(rgt-lft+1-max_cnt>k){
                    help_map[s[lft]]--;
                    lft++;
                }
                ans=max(ans,rgt-lft+1);
            }
            return ans;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-5.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int ans=0,lft=0,rgt=0;
        HashMap<Character,Integer> map = new HashMap<>();
        while(lft<=rgt&&rgt<s.length()){
            if(map.containsKey(s.charAt(rgt))){
                ans=Math.max(ans,rgt-lft);
                lft=map.get(s.charAt(rgt)) + 1;
                map.clear();
               // for(int i = lft; i < lft; i++) {}
                rgt=lft;
            }
            else {
                map.put(s.charAt(rgt),rgt);
                rgt++;
            }
        }
        ans = Math.max(ans,rgt-lft); // to handle case with all distinct
        return ans;
    }
}
